salt/modules/yumpkg.py

```diff
@@ -650,42 +651,52 @@ def list_pkgs(versions_as_list=False, **kwargs):
             for x in ('removed', 'purge_desired')]):
         return {}
 
-    attr = kwargs.get("attr")
-    if 'pkg.list_pkgs' in __context__:
-        cached = __context__['pkg.list_pkgs']
-        return __salt__['pkg_resource.format_pkg_list'](cached, versions_as_list, attr)
+    attr = kwargs.get('attr')
+    if attr is not None:
+        attr = salt.utils.args.split_input(attr)
 
-    ret = {}
-    cmd = ['rpm', '-qa', '--queryformat',
-           salt.utils.pkg.rpm.QUERYFORMAT.replace('%{REPOID}', '(none)') + '\n']
-    output = __salt__['cmd.run'](cmd,
-                                 python_shell=False,
-                                 output_loglevel='trace')
-    for line in output.splitlines():
-        pkginfo = salt.utils.pkg.rpm.parse_pkginfo(
-            line,
-            osarch=__grains__['osarch']
-        )
-        if pkginfo is not None:
-            # see rpm version string rules available at https://goo.gl/UGKPNd
-            pkgver = pkginfo.version
-            epoch = ''
-            release = ''
-            if ':' in pkgver:
-                epoch, pkgver = pkgver.split(":", 1)
-            if '-' in pkgver:
-                pkgver, release = pkgver.split("-", 1)
-            all_attr = {'epoch': epoch, 'version': pkgver, 'release': release,
-                        'arch': pkginfo.arch, 'install_date': pkginfo.install_date,
-                        'install_date_time_t': pkginfo.install_date_time_t}
-            __salt__['pkg_resource.add_pkg'](ret, pkginfo.name, all_attr)
-
-    for pkgname in ret:
-        ret[pkgname] = sorted(ret[pkgname], key=lambda d: d['version'])
-
-    __context__['pkg.list_pkgs'] = ret
+    contextkey = 'pkg.list_pkgs'
 
-    return __salt__['pkg_resource.format_pkg_list'](ret, versions_as_list, attr)
+    if contextkey not in __context__:
```

Comment:
If I am reading this correctly, the `__context__[contextkey]` will be read only once per runtime and `rpm -qa` will _always_ return cached outdated values. Or I just missed something?..

Reply:
This change is functionally equivalent to how we did things before, it just changes how we return data from the function. Any time packages are modified in Salt we clear that key from the `__context__` dunder (look for all the places we do `__context__.pop('pkg.list_pkgs', None)`).

Reply:
Ah. That exactly the part I missed. Then it makes sense, thanks!